feat(deployer/eksctl): extend CLI options, support config file inputs

ytsssun opened the pull request:
*Issue #, if available:* 
#595 
*Description of changes:*

This PR enhances the `aws-k8s-tester` eksctl deployer by adding extensive configuration options. Highlights:
1. Supports config file directly for eksctl.
2. Supports deploying unmanaged nodegroup.
3. Supports deploying **just** the nodegroup.

### Added Options/Flags
- `--config-file`: Path to eksctl config file (if provided, other flags are ignored)
- `--cluster-name`: Name of the EKS cluster (defaults to RunID if not specified)
- `--availability-zones`: Node availability zones
- `--ami-family`: AMI family to use (AmazonLinux2, Bottlerocket)
- `--efa-enabled`: Enable Elastic Fabric Adapter for the nodegroup
- `--volume-size`: Size of the node root volume in GB
- `--private-networking`: Use private networking for nodes
- `--with-oidc`: Enable OIDC provider for IAM roles for service accounts
- `--skip-cluster-creation`: Skip cluster creation, only create nodegroups
- `--unmanaged-nodegroup`: Use unmanaged nodegroup instead of managed nodegroup
- `--nodegroup-name`: Name of the nodegroup

### Test done
1. Test the new flags introduced
```
docker run --rm \   
  --env-file <(env | grep AWS) \      
  -it kubetest2 \                                                               
  kubetest2 eksctl \                                                            
  --cluster-name=my-cluster \                              
  --nodegroup-name=my-ng3 \
  --region=us-west-2 \              
  --kubernetes-version=1.25 \
  --with-oidc \
  --ami-family=Bottlerocket \
  --ami=xxx \
  --instance-types=g5.2xlarge \
  --nodes=1 \
  --volume-size=400 \
  --private-networking \
  --skip-cluster-creation \
  --up \
  --test=exec -- /bin/nvidia.test \
  --test.timeout=60m \
  --test.v \
  --installDevicePlugin=false \
  --feature="unit-test" \
  -efaEnabled=false \
  -nvidiaTestImage=$IMAGE
```

2. Test the config file
```
docker run --rm \                                                                                                                                   
  --env-file <(env | grep AWS) \
  -v ./test-eks-cluster.yaml:/config/test-eks-cluster.yaml \
  -it kubetest2 \
  kubetest2 eksctl \
  --cluster-name=my-cluster \
  --config-file=/config/test-eks-cluster.yaml \
  --region=us-west-2 \
  --skip-cluster-creation \
  --up \
  --test=exec -- /bin/nvidia.test \
  --test.timeout=60m \
  --test.v \
  --installDevicePlugin=false \
  --feature="unit-test" \
  -efaEnabled=false \
  -nvidiaTestImage=$IMAGE
``` 
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Issacwww commented on internal/deployers/eksctl/up.go:
```diff
@@ -13,74 +14,174 @@ import (
+
+	if d.ConfigFile != "" {
+		// If config file is provided, use it
+		if d.SkipClusterCreation {
+			klog.Infof("Adding nodegroup to existing cluster %s using config file: %s", d.clusterName, d.ConfigFile)
+			args = []string{
+				"create",
+				"nodegroup",
+				"--config-file", d.ConfigFile,
+			}
+		} else {
+			klog.Infof("Creating cluster with config file: %s", d.ConfigFile)
+			args = []string{
+				"create",
+				"cluster",
+				"--config-file", d.ConfigFile,
+			}
+		}
```
this seems duplicate with below, can moved outside of configFile check

Issacwww reviewed (commented):
Thanks @ytsssun for contribution! 

FYI, we have this [eksapi-janitor](https://github.com/aws/aws-k8s-tester/tree/main/cmd/kubetest2-eksapi-janitor) for situation that cluster creation failed due to resource limit and need a clean up. You can add similar one if this fit your use case

ytsssun commented:
> FYI, we have this [eksapi-janitor](https://github.com/aws/aws-k8s-tester/tree/main/cmd/kubetest2-eksapi-janitor) for situation that cluster creation failed due to resource limit and need a clean up. You can add similar one if this fit your use case

Thanks, I believe eksctl supports resource deletion via "eksctl delete cluster -f xx.yaml" or even just point to the cluster name. We can add a thing wrapper to that, but I think it would be not different than running just the eksctl commands.

ytsssun commented:
Addressed comments.

Minor refactoring to honor the cluster name from the cluster config so that we don't need to specify the "--cluster-name" when using the config file directly.

Issacwww reviewed (approved):
LGTM, Thanks! 🚀 @ytsssun

cartermckinnon commented on internal/deployers/eksctl/up.go:
```diff
@@ -13,74 +14,164 @@ import (
-		klog.V(2).Infof("Using default number of nodes: %d", d.Nodes)
+
+	// If Bottlerocket AMI family is specified with a custom AMI ID,
+	// ensure we use unmanaged nodegroups as managed nodegroups don't support this combination
```
this should be supported, you'd just need to provide the proper user data in your launch template. Or is this an eksctl limitation?

ytsssun replied:
yes, this is a limitation on eksctl. It does not support custom Bottlerocket AMI Id when using MNG.

mselim00 replied:
IIUC eksctl does _kind of_ support this https://eksctl.io/usage/custom-ami-support/#bottlerocket-custom-ami-support.

I think this is an area where the config file option really shines, when trying to get extra functionality like this, we wouldn't have to add extra logic in this deployer for quirks of the eksctl schema that might change over time.

ytsssun replied:
Like I mentioned, for BR it is supported when using nodegroup. But not for ManagedNodeGroup. `eksctl` would just give failure.

ytsssun commented on internal/deployers/eksctl/deployer.go:
```diff
@@ -24,6 +27,8 @@ type deployer struct {
 	awsConfig      aws.Config
 	eksClient      *eks.Client
 	KubeconfigPath string `flag:"kubeconfig" desc:"Path to kubeconfig"`
+	// ClusterName is the effective cluster name (from flag or RunID)
+	clusterName string
```
Is there a particular reason why run-id is preferred? Is it because we always want a clean end-to-end run (including cluster creation and destruction)?

ytsssun commented on internal/deployers/eksctl/up.go:
```diff
@@ -13,74 +14,164 @@ import (
+	VolumeSize            int      `flag:"volume-size" desc:"Size of the node root volume in GB"`
+	PrivateNetworking     bool     `flag:"private-networking" desc:"Use private networking for nodes"`
+	WithOIDC              bool     `flag:"with-oidc" desc:"Enable OIDC provider for IAM roles for service accounts"`
+	SkipClusterCreation   bool     `flag:"skip-cluster-creation" desc:"Skip cluster creation, only create nodegroups"`
```
Thank you for your feedback. Let me clarify the rationale behind the `--skip-cluster-creation` option:

While eksctl does have checking capabilities for nodegroups and clusters, it fails with an error when attempting `eksctl create cluster` on an existing cluster:

```
2025-03-26 20:33:44 [✖]  creating CloudFormation stack "eksctl-efa-p4-cluster-125-cluster": operation error CloudFormation: CreateStack, https response error StatusCode: 400, RequestID: 3239e519-0c64-47cc-a993-efd6ca6205bc, AlreadyExistsException: Stack [eksctl-efa-p4-cluster-125-cluster] already exists
Error: failed to create cluster "efa-p4-cluster-125"
```

This feature addresses specific testing workflows on our side:

1. **Testing efficiency**: Cluster creation is time-consuming. When iterating on nodegroup configurations or troubleshooting nodegroup deployment issues, redeploying the entire cluster is inefficient.

2. **Intermediate approach**: This isn't about static clusters (which does not work for us as it assumes the cluster and nodegroups are ready, in which case it does not add much value comparing to just run the `go test` command directly). Rather, it allows controlled management of the cluster lifecycle separately from nodegroups.

3. **Test isolation**: We often need to deploy different nodegroup configurations on the same cluster for different features without recreating the entire cluster.

The option provides a middle ground between fully ephemeral clusters and static clusters, enabling more flexible testing patterns while still maintaining the ability to automate the full deployment process when needed.

ytsssun replied:
> Wondering if we can just pass it along more directly, e.g. the flag can be --create and the value of it is the argument to the eksctl create call, defaulting to cluster.

That is definitely a valid point. But it further extended the scope of the PR. Also, as eksctl deployer, being able to redeploy the nodegroup is definitely a more "significant" need comparing to other options listed there. I don't see a compelling reason to must include all the "eksctl create" subcommands but it does extend the functionality. 

I would still prefer to focus on support nodegroup creation and track others in a separate effort.

ytsssun commented on internal/deployers/eksctl/deployer.go:
```diff
@@ -66,5 +71,59 @@ func bindFlags(d *deployer) *pflag.FlagSet {
+}
+
+// parseClusterNameFromConfig extracts the cluster name from an eksctl config file
+func (d *deployer) parseClusterNameFromConfig(configFilePath string) (string, error) {
```
Thanks for the feedback. I'd like to clarify my reasoning for adding the `--cluster-name` flag:

1. Sematically, using `--cluster-name` is more intuitive than overriding `--run-id`, which may affect components beyond just the cluster name.

2. As I mentioned in our use cases, we sometimes would like to add nodegroups to existing clusters, so run-id will not work for us.

3. The implementation still defaults to using `runID` when `--cluster-name` isn't specified, maintaining the clean run pattern by default.

For config files, I believe we should respect the user's intent when they explicitly provide configuration. If strong naming control is needed, we could consider a separate flag to enforce `runID` as the cluster name.

What specific concerns do you have about allowing custom cluster names as an option?

ytsssun commented on internal/deployers/eksctl/cluster_config.go:
```diff
@@ -16,29 +16,92 @@ metadata:
+      source /var/lib/cloud/scripts/eksctl/bootstrap.helper.sh
+      /etc/eks/bootstrap.sh {{.ClusterName}} --kubelet-extra-args "--node-labels=${NODE_LABELS}"
+    {{- end}}
+{{- else}}
```
I updated the logic in here to directly using upstream [ClusterConfig](https://pkg.go.dev/github.com/weaveworks/eksctl@v0.208.0/pkg/apis/eksctl.io/v1alpha5#ClusterConfig).

ytsssun commented on internal/deployers/eksctl/up.go:
```diff
@@ -13,14 +14,31 @@ import (
+	InstanceTypes         []string `flag:"instance-types" desc:"Node instance types"`
+	ConfigFile            string   `flag:"config-file" desc:"Path to eksctl config file (if provided, other flags are ignored)"`
+	AvailabilityZones     []string `flag:"availability-zones" desc:"Node availability zones"`
+	AMIFamily             string   `flag:"ami-family" desc:"AMI family to use (AmazonLinux2, Bottlerocket)"`
```
Updated.

ytsssun commented on internal/deployers/eksctl/up.go:
```diff
@@ -30,57 +48,130 @@ func (d *deployer) verifyUpFlags() error {
 	}
-	if d.Nodes < 0 {
-		return fmt.Errorf("number of nodes must be greater than zero")
+	if d.Nodes <= 0 {
+		return fmt.Errorf("number of nodes must be equal or greater than zero")
 	}
-	if d.Nodes == 0 {
-		d.Nodes = 4
-		klog.V(2).Infof("Using default number of nodes: %d", d.Nodes)
```
Not really any motivation, I can keep the default value of 4 here for consistency.

ytsssun commented:
Adding newest testing result I have:

```
> docker run --rm \
                                  -e AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
                                  -e AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY \
                                  -e AWS_SESSION_TOKEN=$AWS_SESSION_TOKEN \
                                  -it kubetest2 \
                                  kubetest2 eksctl \
                                  --cluster-name=my-cluster-130 \
                                  --nodegroup-name=my-ng2 \
                                  --region=us-west-2 \
                                  --kubernetes-version=1.30 \
                                  --with-oidc \
                                  --ami-family=Bottlerocket \
                                  --ami=xxx \
                                  --instance-types=g5.2xlarge \
                                  --nodes=1 \
                                  --volume-size=400 \
                                  --private-networking \
                                  --skip-cluster-creation \
                                  --up \
                                  --test=exec -- /bin/nvidia.test \
                                  --test.timeout=60m \
                                  --test.v \
                                  --installDevicePlugin=false \
                                  --feature="unit-test" \
                                  -efaEnabled=false \
                                  -nvidiaTestImage=xxxx.dkr.ecr.us-west-2.amazonaws.com/nccl-test:nvidia-v1-amd
I0528 22:23:46.570969      13 app.go:61] The files in RunDir shall not be part of Artifacts
I0528 22:23:46.571045      13 app.go:62] pass rundir-in-artifacts flag True for RunDir to be part of Artifacts
I0528 22:23:46.571051      13 app.go:64] RunDir for this run: "/workdir/_rundir/f0559119-1547-4ed0-8f63-343a842debbb"
I0528 22:23:46.573479      13 app.go:136] ID for this run: "xxx"
W0528 22:23:46.573491      13 up.go:58] Bottlerocket with custom AMI requires unmanaged nodegroups. Setting --unmanaged-nodegroup=true
I0528 22:23:46.573496      13 up.go:84] Using unmanaged nodegroup for cluster my-cluster-130
I0528 22:23:46.573504      13 cluster_config.go:100] rendering cluster config yaml based on the ClusterConfig: &TypeMeta{Kind:ClusterConfig,APIVersion:eksctl.io/v1alpha5,}
I0528 22:23:46.574468      13 up.go:100] Rendered cluster config: accessConfig: {}
addonsConfig: {}
apiVersion: eksctl.io/v1alpha5
cloudWatch:
  clusterLogging: {}
iam:
  withOIDC: true
kind: ClusterConfig
kubernetesNetworkConfig:
  ipFamily: IPv4
metadata:
  name: my-cluster-130
  region: us-west-2
  version: "1.30"
nodeGroups:
- amiFamily: Bottlerocket
  desiredCapacity: 1
  disableIMDSv1: true
  disablePodIMDS: false
  efaEnabled: false
  iam:
    withAddonPolicies:
      albIngress: false
      appMesh: false
      appMeshPreview: false
      autoScaler: false
      awsLoadBalancerController: false
      certManager: false
      cloudWatch: false
      ebs: false
      efs: false
      externalDNS: false
      fsx: false
      imageBuilder: false
      xRay: false
  instanceSelector: {}
  instanceType: g5.2xlarge
  maxSize: 1
  minSize: 1
  name: my-ng2
  privateNetworking: true
  securityGroups:
    withLocal: true
    withShared: true
  ssh:
    allow: false
    publicKeyPath: ~/.ssh/id_rsa.pub
  volumeSize: 400
  volumeType: gp3
privateCluster:
  enabled: false
  skipEndpointCreation: false
vpc:
  autoAllocateIPv6: false
  cidr: 192.168.0.0/16
  manageSharedNodeSecurityGroupRules: true
  nat:
    gateway: Single
I0528 22:23:46.577031      13 up.go:154] Adding nodegroup to existing cluster my-cluster-130
2025-05-28 22:23:46 [!]  Amazon EKS will no longer publish EKS-optimized Amazon Linux 2 (AL2) AMIs after November 26th, 2025. Additionally, Kubernetes version 1.32 is the last version for which Amazon EKS will release AL2 AMIs. From version 1.33 onwards, Amazon EKS will continue to release AL2023 and Bottlerocket based AMIs. The default AMI family when creating clusters and nodegroups in Eksctl will be changed to AL2023 in the future.
2025-05-28 22:23:47 [ℹ]  nodegroup "my-ng2" will use "ami-002d165d9f0a9cd14" [Bottlerocket/1.30]
2025-05-28 22:23:48 [ℹ]  1 existing nodegroup(s) (TachyonEFA-v02) will be excluded
2025-05-28 22:23:48 [ℹ]  1 nodegroup (my-ng2) was included (based on the include/exclude rules)
2025-05-28 22:23:48 [ℹ]  will create a CloudFormation stack for each of 1 nodegroups in cluster "my-cluster-130"
2025-05-28 22:23:48 [ℹ]
2 sequential tasks: { fix cluster compatibility, 1 task: { 1 task: { create nodegroup "my-ng2" } }
}
2025-05-28 22:23:48 [ℹ]  checking cluster stack for missing resources
2025-05-28 22:23:48 [ℹ]  cluster stack has all required resources
2025-05-28 22:23:48 [ℹ]  building nodegroup stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:23:48 [ℹ]  skipping us-west-2d from selection because it doesn't support the following instance type(s): g5.2xlarge
2025-05-28 22:23:48 [ℹ]  deploying stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:23:48 [ℹ]  waiting for CloudFormation stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:24:19 [ℹ]  waiting for CloudFormation stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:24:56 [ℹ]  waiting for CloudFormation stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:25:38 [ℹ]  waiting for CloudFormation stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:27:06 [ℹ]  waiting for CloudFormation stack "eksctl-my-cluster-130-nodegroup-my-ng2"
2025-05-28 22:27:06 [ℹ]  no tasks
2025-05-28 22:27:06 [ℹ]  nodegroup "my-ng2" has 1 node(s)
2025-05-28 22:27:06 [ℹ]  node "xxx.us-west-2.compute.internal" is ready
2025-05-28 22:27:06 [ℹ]  waiting for at least 1 node(s) to become ready in "my-ng2"
2025-05-28 22:27:06 [ℹ]  nodegroup "my-ng2" has 1 node(s)
2025-05-28 22:27:06 [ℹ]  node "ixxx.us-west-2.compute.internal" is ready
2025-05-28 22:27:06 [✔]  created 1 nodegroup(s) in cluster "my-cluster-130"
2025-05-28 22:27:06 [✔]  created 0 managed nodegroup(s) in cluster "my-cluster-130"
2025-05-28 22:27:06 [ℹ]  checking security group configuration for all nodegroups
2025-05-28 22:27:06 [ℹ]  all nodegroups have up-to-date cloudformation templates
I0528 22:27:06.693324      13 up.go:133] Writing kubeconfig to /workdir/_rundir/f0559119-1547-4ed0-8f63-343a842debbb/kubeconfig
2025-05-28 22:27:07 [✔]  saved kubeconfig as "/workdir/_rundir/f0559119-1547-4ed0-8f63-343a842debbb/kubeconfig"
I0528 22:27:07.429612      13 up.go:147] Successfully wrote kubeconfig to /workdir/_rundir/f0559119-1547-4ed0-8f63-343a842debbb/kubeconfig
=== RUN   TestContainerdConfig
=== RUN   TestContainerdConfig/containerd-config-check
    env.go:438: Skipping feature "containerd-config-check": name not matched
--- PASS: TestContainerdConfig (0.00s)
    --- SKIP: TestContainerdConfig/containerd-config-check (0.00s)
=== RUN   TestMPIJobPytorchTraining
=== RUN   TestMPIJobPytorchTraining/single-node
    env.go:438: Skipping feature "single-node": name not matched
=== RUN   TestMPIJobPytorchTraining/multi-node
    env.go:438: Skipping feature "multi-node": name not matched
2025/05/28 22:27:13 No node type specified. Using the node type g5.2xlarge in the node groups.
--- PASS: TestMPIJobPytorchTraining (0.00s)
    --- SKIP: TestMPIJobPytorchTraining/single-node (0.00s)
    --- SKIP: TestMPIJobPytorchTraining/multi-node (0.00s)
=== RUN   TestSingleNodeUnitTest
=== RUN   TestSingleNodeUnitTest/unit-test
=== RUN   TestSingleNodeUnitTest/unit-test/Unit_test_Job_succeeds
=== NAME  TestSingleNodeUnitTest/unit-test
    unit_test.go:82: Test log for unit-test-job:
    unit_test.go:83: # Running tests in gpu_unit_tests/tests/test_basic.sh
        ok - test_01_device_query
        ok - test_02_vector_add
        ok - test_03_bandwidth
        ok - test_04_dcgm_diagnostics
        # Running tests in gpu_unit_tests/tests/test_sysinfo.sh
        diff: test_sysinfo.sh.data/g5.2xlarge/numa_topo.txt: No such file or directory
        ok - test_numa_topo_topo
        diff: test_sysinfo.sh.data/g5.2xlarge/gpu_count.txt: No such file or directory
        ok - test_nvidia_gpu_count
        ok - test_nvidia_gpu_throttled
        ok - test_nvidia_gpu_unused
        diff: test_sysinfo.sh.data/g5.2xlarge/nvidia_persistence_status.txt: No such file or directory
        ok - test_nvidia_persistence_status
        diff: test_sysinfo.sh.data/g5.2xlarge/nvidia_smi_topo.txt: No such file or directory
        ok - test_nvidia_smi_topo

=== RUN   TestSingleNodeUnitTest/hpc-benckmarks
    env.go:438: Skipping feature "hpc-benckmarks": name not matched
--- PASS: TestSingleNodeUnitTest (180.08s)
    --- PASS: TestSingleNodeUnitTest/unit-test (180.08s)
        --- PASS: TestSingleNodeUnitTest/unit-test/Unit_test_Job_succeeds (180.01s)
    --- SKIP: TestSingleNodeUnitTest/hpc-benckmarks (0.00s)
PASS
```

ytsssun commented on internal/deployers/eksctl/up.go:
```diff
@@ -30,57 +48,130 @@ func (d *deployer) verifyUpFlags() error {
+		}
+	}
+}
+
```
Addressed the comments. Using `DeployTarget` as the input.

mselim00 reviewed (approved):
lgtm, thanks for the improvements!
